coderd/inboxnotifications.go

```diff
@@ -0,0 +1,376 @@
+	return codersdk.InboxNotification{
+		ID:         notif.ID,
+		UserID:     notif.UserID,
+		TemplateID: notif.TemplateID,
+		Targets:    notif.Targets,
+		Title:      notif.Title,
+		Content:    notif.Content,
+		Icon:       notif.Icon,
+		Actions: func() []codersdk.InboxNotificationAction {
+			var actionsList []codersdk.InboxNotificationAction
+			err := json.Unmarshal([]byte(notif.Actions), &actionsList)
+			if err != nil {
+				logger.Error(ctx, "unmarshal inbox notification actions", slog.Error(err))
+			}
+			return actionsList
+		}(),
+		ReadAt: func() *time.Time {
+			if !notif.ReadAt.Valid {
+				return nil
+			}
+			return &notif.ReadAt.Time
+		}(),
+		CreatedAt: notif.CreatedAt,
+	}
+}
+
+// watchInboxNotifications watches for new inbox notifications and sends them to the client.
+// The client can specify a list of target IDs to filter the notifications.
+// @Summary Watch for new inbox notifications
+// @ID watch-for-new-inbox-notifications
+// @Security CoderSessionToken
+// @Produce json
+// @Tags Notifications
+// @Param targets query string false "Comma-separated list of target IDs to filter notifications"
+// @Param templates query string false "Comma-separated list of template IDs to filter notifications"
+// @Param read_status query string false "Filter notifications by read status. Possible values: read, unread, all"
+// @Success 200 {object} codersdk.GetInboxNotificationResponse
+// @Router /notifications/inbox/watch [get]
+func (api *API) watchInboxNotifications(rw http.ResponseWriter, r *http.Request) {
+	ctx := r.Context()
+
+	var (
+		apikey          = httpmw.APIKey(r)
+		targetsParam    = r.URL.Query().Get("targets")
+		templatesParam  = r.URL.Query().Get("templates")
+		readStatusParam = r.URL.Query().Get("read_status")
+	)
+
+	targets, templates, readStatusParam, err := convertInboxNotificationParameters(ctx, api.Logger, targetsParam, templatesParam, readStatusParam)
+	if err != nil {
+		httpapi.Write(ctx, rw, http.StatusBadRequest, codersdk.Response{
+			Message: "Invalid query parameter.",
+			Detail:  err.Error(),
+		})
+		return
+	}
+
+	conn, err := websocket.Accept(rw, r, nil)
+	if err != nil {
+		httpapi.Write(ctx, rw, http.StatusInternalServerError, codersdk.Response{
+			Message: "Failed to upgrade connection to websocket.",
+			Detail:  err.Error(),
+		})
+		return
+	}
+
+	go httpapi.Heartbeat(ctx, conn)
+	defer conn.Close(websocket.StatusNormalClosure, "connection closed")
+
+	notificationCh := make(chan codersdk.InboxNotification, 1)
+
+	closeInboxNotificationsSubscriber, err := api.Pubsub.SubscribeWithErr(pubsub.InboxNotificationForOwnerEventChannel(apikey.UserID),
+		pubsub.HandleInboxNotificationEvent(
+			func(ctx context.Context, payload pubsub.InboxNotificationEvent, err error) {
+				if err != nil {
+					api.Logger.Error(ctx, "inbox notification event", slog.Error(err))
+					return
+				}
+
+				// filter out notifications that don't match the targets
```

Comment:
Curious why we need this here when the query allows a list of targets to match on?

Reply:
+1

unless this is intended, then please drop a comment on why this is mandatory.